### PR TITLE
Introduce string_view and use it for gpr_split_host_port

### PR DESCRIPTION
**Pull request by soheilhy**
This is a roll forward of #19218 with a fix.

The issue was that instead of using the return value,
the original patch was checking the length of the host
and port.

The fix is as simple as the following two-line change.
Added unittests in `test/core/gprpp/host_port_test.cc` to
resolve the issue.

```
diff --git a/src/core/lib/gprpp/host_port.cc b/src/core/lib/gprpp/host_port.cc
index f3f8a73602..bb7ea5c391 100644
--- a/src/core/lib/gprpp/host_port.cc
+++ b/src/core/lib/gprpp/host_port.cc
@@ -90,8 +90,8 @@ bool SplitHostPort(StringView name, UniquePtr<char>* host,
   StringView host_view;
   StringView port_view;
   const bool ret = SplitHostPort(name, &host_view, &port_view);
-  host->reset(host_view.empty() ? nullptr : host_view.dup().release());
-  port->reset(port_view.empty() ? nullptr : port_view.dup().release());
+  host->reset(ret ? host_view.dup().release() : nullptr);
+  port->reset(ret ? port_view.dup().release() : nullptr);
   return ret;
 }
 }  // namespace grpc_core
```

### CODE REVIEW

File: src/core/lib/gprpp/host_port.cc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  }
+  return true;
+}
+
+bool SplitHostPort(StringView name, UniquePtr<char>* host,
+                   UniquePtr<char>* port) {
+  StringView host_view;
+  StringView port_view;
+  const bool ret = SplitHostPort(name, &host_view, &port_view);
+  host->reset(ret ? host_view.dup().release() : nullptr);
```
Comment (markdroth) [inline src/core/lib/gprpp/host_port.cc]: This might be better written as:

```
if (ret) {
  *host = host_view.dup();
  *port = port_view.dup();
}
```

Comment (soheilhy) [reply src/core/lib/gprpp/host_port.cc]: That's a great idea. I also got rid of the `reset(nullptr)` path by adding debug asserts above.

**Comment by soheilhy** [comment]
Sorry about that @markdroth. I split the patch in two: first is the clean revert, and second is the fix. thanks for the review!

**Comment by soheilhy** [comment]
@markdroth sorry about the hassle, but I had to make two changes and rebase everything.

1) I had to fix the `credential_test.cc` and revert in this PR.
2) I had to fix the fix.  Problem is that the code always returns a `host` string when it's able to split, but `port` will be `nullptr` if it's empty regardless of the return value.

I pushed the extra commits above,  and will wait to see if there are other cases that I've missed. Sorry about the extra round-trip.

**Comment by markdroth** [comment]
Still LGTM.

**Comment by soheilhy** [comment]
Thank you! Will request an internal cherrypick then.

**Comment by soheilhy** [comment]
Merging for cherry-pick.

**Pull request merged**